bin/propolis-server/src/lib/vm/state_driver.rs

```diff
@@ -524,32 +535,208 @@ impl StateDriver {
-                                 "error" => ?e);
-                self.external_state.update(ExternalStateUpdate::Instance(
-                    InstanceState::Failed,
-                ));
+        }
+
+        // Next, prepare to start block backends. This is done by capturing the
+        // current block backend set and creating a future that issues all the
+        // start requests.
+        //
+        // For this to work, the set of block backends to be started must not
+        // change while the VM is starting. This is guaranteed because all such
+        // requests to hotplug a block backend will be dispatched to the VM's
+        // request queue; if any such requests are seen below, they can simply
+        // be buffered and handled after the rest of the VM has started.
+        async fn start_block_backends(
+            log: slog::Logger,
+            backends: BlockBackendMap,
+        ) -> anyhow::Result<()> {
+            for (name, backend) in backends {
+                info!(log, "starting block backend {}", name);
+                let res = backend.start().await;
+                if let Err(e) = &res {
+                    error!(
+                        log,
+                        "startup failed for block backend {}: {:?}", name, e
+                    );
+
+                    return res;
+                }
             }
+
+            Ok(())
         }
 
-        start_result
+        let block_backends = objects.block_backend_map().clone();
+        let block_backend_fut =
+            start_block_backends(self.log.clone(), block_backends);
+        tokio::pin!(block_backend_fut);
+
+        // Drop the VM object lock before proceeding to allow other API calls
+        // that simply read the VM to make progress. Again, note that the set of
+        // objects being started still can't change, not because the lock is
+        // held, but because the only entity that can change them is the current
+        // task, which can decide whether and how to buffer incoming requests.
+        drop(objects);
+
+        // Keep track of whether the external queue produced a request to stop
+        // the VM while it was being started. If such a request is seen, send a
+        // self-request to stop just before returning so that the VM will stop
+        // immediately.
+        enum Selection {
+            BackendFuture(anyhow::Result<()>),
+            Event(InputQueueEvent),
+        }
+        loop {
+            let selection = tokio::select! {
+                // If the VM successfully starts, return immediately and let
+                // the caller process any events that may happen to be on the
+                // queue.
+                biased;
+
+                res = &mut block_backend_fut => {
+                    Selection::BackendFuture(res)
+                }
+
+                event = self.input_queue.wait_for_next_event() => {
+                    Selection::Event(event)
+                }
```

Comment:
`wait_for_next_event()` won't actually remove an event from the queue unless it completes, right? I assume this is the case but wanted to make extra sure that when block backend startup completes, we won't consume an event from the queue that _should_ still be there for the caller to consume once the VM is started.

Reply:
Correct, `wait_for_next_event` should be `select!`-safe: if it pops something from the queue, then it always returns the popped value immediately without awaiting again.